// @datadog/mobile-react-native/lib/commonjs/rum/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PropagatorType = exports.ErrorSource = exports.RumActionType = void 0;

/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

/**
 * The entry point to use Datadog's RUM feature.
 */

/**
 * Describe the type of a RUM Action.
 */
let RumActionType;
exports.RumActionType = RumActionType;

(function (RumActionType) {
  RumActionType["TAP"] = "TAP";
  RumActionType["SCROLL"] = "SCROLL";
  RumActionType["SWIPE"] = "SWIPE";
  RumActionType["BACK"] = "BACK";
  RumActionType["CUSTOM"] = "CUSTOM";
})(RumActionType || (exports.RumActionType = RumActionType = {}));

let ErrorSource;
/**
 * Type of instrumentation on the host.
 * - DATADOG: Datadog’s propagator (`x-datadog-*`)
 * - TRACECONTEXT: W3C Trace Context (`traceparent`)
 * - B3: B3 single header (`b3`)
 * - B3MULTI: B3 multiple headers (`X-B3-*`)
 */

exports.ErrorSource = ErrorSource;

(function (ErrorSource) {
  ErrorSource["NETWORK"] = "NETWORK";
  ErrorSource["SOURCE"] = "SOURCE";
  ErrorSource["CONSOLE"] = "CONSOLE";
  ErrorSource["WEBVIEW"] = "WEBVIEW";
  ErrorSource["CUSTOM"] = "CUSTOM";
})(ErrorSource || (exports.ErrorSource = ErrorSource = {}));

let PropagatorType;
exports.PropagatorType = PropagatorType;

(function (PropagatorType) {
  PropagatorType["DATADOG"] = "datadog";
  PropagatorType["TRACECONTEXT"] = "tracecontext";
  PropagatorType["B3"] = "b3";
  PropagatorType["B3MULTI"] = "b3multi";
})(PropagatorType || (exports.PropagatorType = PropagatorType = {}));
