// @datadog/mobile-react-native/lib/module/rum/types.js
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

/**
 * The entry point to use Datadog's RUM feature.
 */

/**
 * Describe the type of a RUM Action.
 */
export let RumActionType;

(function (RumActionType) {
  RumActionType["TAP"] = "TAP";
  RumActionType["SCROLL"] = "SCROLL";
  RumActionType["SWIPE"] = "SWIPE";
  RumActionType["BACK"] = "BACK";
  RumActionType["CUSTOM"] = "CUSTOM";
})(RumActionType || (RumActionType = {}));

export let ErrorSource;
/**
 * Type of instrumentation on the host.
 * - DATADOG: Datadog’s propagator (`x-datadog-*`)
 * - TRACECONTEXT: W3C Trace Context (`traceparent`)
 * - B3: B3 single header (`b3`)
 * - B3MULTI: B3 multiple headers (`X-B3-*`)
 */

(function (ErrorSource) {
  ErrorSource["NETWORK"] = "NETWORK";
  ErrorSource["SOURCE"] = "SOURCE";
  ErrorSource["CONSOLE"] = "CONSOLE";
  ErrorSource["WEBVIEW"] = "WEBVIEW";
  ErrorSource["CUSTOM"] = "CUSTOM";
})(ErrorSource || (ErrorSource = {}));

export let PropagatorType;

(function (PropagatorType) {
  PropagatorType["DATADOG"] = "datadog";
  PropagatorType["TRACECONTEXT"] = "tracecontext";
  PropagatorType["B3"] = "b3";
  PropagatorType["B3MULTI"] = "b3multi";
})(PropagatorType || (PropagatorType = {}));
