// @datadog/mobile-react-native/lib/commonjs/DdSdkReactNative.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DdSdkReactNative = void 0;

var _reactNative = require("react-native");

var _DdSdkReactNativeConfiguration = require("./DdSdkReactNativeConfiguration");

var _InternalLog = require("./InternalLog");

var _ProxyConfiguration = require("./ProxyConfiguration");

var _SdkVerbosity = require("./SdkVerbosity");

var _foundation = require("./foundation");

var _DdLogs = require("./logs/DdLogs");

var _longTasksUtils = require("./longTasksUtils");

var _DdRum = require("./rum/DdRum");

var _DdRumErrorTracking = require("./rum/instrumentation/DdRumErrorTracking");

var _DdRumUserInteractionTracking = require("./rum/instrumentation/interactionTracking/DdRumUserInteractionTracking");

var _DdRumResourceTracking = require("./rum/instrumentation/resourceTracking/DdRumResourceTracking");

var _AttributesSingleton = require("./sdk/AttributesSingleton/AttributesSingleton");

var _BufferSingleton = require("./sdk/DatadogProvider/Buffer/BufferSingleton");

var _UserInfoSingleton = require("./sdk/UserInfoSingleton/UserInfoSingleton");

var _types = require("./types");

var _version = require("./version");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * This class initializes the Datadog SDK, and sets up communication with the server.
 */
class DdSdkReactNative {
  // Proxy

  /**
   * Initializes the Datadog SDK.
   * @param configuration the configuration for the SDK library
   * @returns a Promise.
   */
  static async initialize(configuration) {
    await DdSdkReactNative.initializeNativeSDK(configuration, {
      initializationModeForTelemetry: 'LEGACY'
    });
    DdSdkReactNative.enableFeatures(configuration);
  }

  /**
   * FOR INTERNAL USE ONLY.
   */
  static async _initializeFromDatadogProvider(configuration) {
    DdSdkReactNative.enableFeatures(configuration);

    if (configuration.initializationMode === _DdSdkReactNativeConfiguration.InitializationMode.SYNC) {
      return DdSdkReactNative.initializeNativeSDK(configuration, {
        initializationModeForTelemetry: 'SYNC'
      });
    }

    if (configuration.initializationMode === _DdSdkReactNativeConfiguration.InitializationMode.ASYNC) {
      return _reactNative.InteractionManager.runAfterInteractions(() => {
        return DdSdkReactNative.initializeNativeSDK(configuration, {
          initializationModeForTelemetry: 'ASYNC'
        });
      });
    } // TODO: Remove when DdSdkReactNativeConfiguration is deprecated


    if (configuration instanceof _DdSdkReactNativeConfiguration.DdSdkReactNativeConfiguration) {
      return DdSdkReactNative.initializeNativeSDK(configuration, {
        initializationModeForTelemetry: 'SYNC'
      });
    }
  }
  /**
   * FOR INTERNAL USE ONLY.
   */


  static async _enableFeaturesFromDatadogProvider(features) {
    DdSdkReactNative.features = features;
    DdSdkReactNative.enableFeatures((0, _DdSdkReactNativeConfiguration.addDefaultValuesToAutoInstrumentationConfiguration)(features));
  }
  /**
   * FOR INTERNAL USE ONLY.
   */


  /**
   * Adds a set of attributes to the global context attached with all future Logs, Spans and RUM events.
   * To remove an attribute, set it to `undefined` in a call to `setAttributes`.
   * @param attributes: The global context attributes.
   * @returns a Promise.
   */
  // eslint-disable-next-line @typescript-eslint/ban-types
  static async setAttributes(attributes) {
    _InternalLog.InternalLog.log(`Setting attributes ${JSON.stringify(attributes)}`, _SdkVerbosity.SdkVerbosity.DEBUG);

    await _foundation.DdSdk.setAttributes(attributes);

    _AttributesSingleton.AttributesSingleton.getInstance().setAttributes(attributes);
  }
  /**
   * Set the user information.
   * @param user: The user object (use builtin attributes: 'id', 'email', 'name', and/or any custom attribute).
   * @returns a Promise.
   */
  // eslint-disable-next-line @typescript-eslint/ban-types


  static async setUser(user) {
    _InternalLog.InternalLog.log(`Setting user ${JSON.stringify(user)}`, _SdkVerbosity.SdkVerbosity.DEBUG);

    await _foundation.DdSdk.setUser(user);

    _UserInfoSingleton.UserInfoSingleton.getInstance().setUserInfo(user);
  }
  /**
   * Set the tracking consent regarding the data collection.
   * @param trackingConsent: One of TrackingConsent values.
   * @returns a Promise.
   */


  static setTrackingConsent(consent) {
    _InternalLog.InternalLog.log(`Setting consent ${consent}`, _SdkVerbosity.SdkVerbosity.DEBUG);

    return _foundation.DdSdk.setTrackingConsent(consent);
  }

  static enableFeatures(configuration) {
    if (DdSdkReactNative.wasAutoInstrumented) {
      _InternalLog.InternalLog.log("Can't auto instrument Datadog, SDK was already instrumented", _SdkVerbosity.SdkVerbosity.WARN);

      return;
    }

    if (configuration.trackInteractions) {
      _DdRumUserInteractionTracking.DdRumUserInteractionTracking.startTracking();
    }

    if (configuration.trackResources) {
      _DdRumResourceTracking.DdRumResourceTracking.startTracking({
        tracingSamplingRate: configuration.resourceTracingSamplingRate,
        firstPartyHosts: (0, _DdSdkReactNativeConfiguration.formatFirstPartyHosts)(configuration.firstPartyHosts)
      });
    }

    if (configuration.trackErrors) {
      _DdRumErrorTracking.DdRumErrorTracking.startTracking();
    }

    if (configuration.logEventMapper) {
      _DdLogs.DdLogs.registerLogEventMapper(configuration.logEventMapper);
    }

    if (configuration.errorEventMapper) {
      _DdRum.DdRum.registerErrorEventMapper(configuration.errorEventMapper);
    }

    if (configuration.resourceEventMapper) {
      _DdRum.DdRum.registerResourceEventMapper(configuration.resourceEventMapper);
    }

    if (configuration.actionEventMapper) {
      _DdRum.DdRum.registerActionEventMapper(configuration.actionEventMapper);
    }

    DdSdkReactNative.wasAutoInstrumented = true;
  }

}

exports.DdSdkReactNative = DdSdkReactNative;

_defineProperty(DdSdkReactNative, "DD_SOURCE_KEY", '_dd.source');

_defineProperty(DdSdkReactNative, "DD_SDK_VERSION", '_dd.sdk_version');

_defineProperty(DdSdkReactNative, "DD_SERVICE_NAME", '_dd.service_name');

_defineProperty(DdSdkReactNative, "DD_SDK_VERBOSITY_KEY", '_dd.sdk_verbosity');

_defineProperty(DdSdkReactNative, "DD_NATIVE_VIEW_TRACKING_KEY", '_dd.native_view_tracking');

_defineProperty(DdSdkReactNative, "DD_NATIVE_INTERACTION_TRACKING_KEY", '_dd.native_interaction_tracking');

_defineProperty(DdSdkReactNative, "DD_VERSION", '_dd.version');

_defineProperty(DdSdkReactNative, "DD_VERSION_SUFFIX", '_dd.version_suffix');

_defineProperty(DdSdkReactNative, "DD_PROXY_TYPE_KEY", '_dd.proxy.type');

_defineProperty(DdSdkReactNative, "DD_PROXY_ADDRESS_KEY", '_dd.proxy.address');

_defineProperty(DdSdkReactNative, "DD_PROXY_PORT_KEY", '_dd.proxy.port');

_defineProperty(DdSdkReactNative, "DD_PROXY_USERNAME_KEY", '_dd.proxy.username');

_defineProperty(DdSdkReactNative, "DD_PROXY_PASSWORD_KEY", '_dd.proxy.password');

_defineProperty(DdSdkReactNative, "wasInitialized", false);

_defineProperty(DdSdkReactNative, "wasAutoInstrumented", false);

_defineProperty(DdSdkReactNative, "features", void 0);

_defineProperty(DdSdkReactNative, "initializeNativeSDK", async (configuration, params) => {
  if (DdSdkReactNative.wasInitialized) {
    _InternalLog.InternalLog.log("Can't initialize Datadog, SDK was already initialized", _SdkVerbosity.SdkVerbosity.WARN);

    if (!__DEV__) {
      _foundation.DdSdk.telemetryDebug('RN SDK was already initialized in javascript');
    }

    return new Promise(resolve => resolve());
  }

  _InternalLog.InternalLog.verbosity = configuration.verbosity;
  DdSdkReactNative.buildConfiguration(configuration);
  await _foundation.DdSdk.initialize(new _types.DdSdkConfiguration(configuration.clientToken, configuration.env, configuration.applicationId, configuration.nativeCrashReportEnabled, (0, _longTasksUtils.adaptLongTaskThreshold)(configuration.nativeLongTaskThresholdMs), (0, _longTasksUtils.adaptLongTaskThreshold)(configuration.longTaskThresholdMs), configuration.sampleRate === undefined ? configuration.sessionSamplingRate : configuration.sampleRate, configuration.site, configuration.trackingConsent, configuration.additionalConfig, configuration.telemetrySampleRate, configuration.vitalsUpdateFrequency, {
    initializationType: params.initializationModeForTelemetry,
    trackErrors: configuration.trackErrors,
    trackInteractions: configuration.trackInteractions,
    trackNetworkRequests: configuration.trackResources
  }));

  _InternalLog.InternalLog.log('Datadog SDK was initialized', _SdkVerbosity.SdkVerbosity.INFO);

  DdSdkReactNative.wasInitialized = true;

  _BufferSingleton.BufferSingleton.onInitialization();
});

_defineProperty(DdSdkReactNative, "_initializeFromDatadogProviderWithConfigurationAsync", async configuration => {
  if (!DdSdkReactNative.features) {
    _InternalLog.InternalLog.log("Can't initialize Datadog, make sure the DatadogProvider component is mounted before calling this function", _SdkVerbosity.SdkVerbosity.WARN);

    return new Promise(resolve => resolve());
  }

  return DdSdkReactNative.initializeNativeSDK((0, _DdSdkReactNativeConfiguration.buildConfigurationFromPartialConfiguration)(DdSdkReactNative.features, configuration), {
    initializationModeForTelemetry: 'PARTIAL'
  });
});

_defineProperty(DdSdkReactNative, "buildConfiguration", configuration => {
  configuration.additionalConfig[DdSdkReactNative.DD_SOURCE_KEY] = 'react-native';
  configuration.additionalConfig[DdSdkReactNative.DD_SDK_VERSION] = _version.version;
  configuration.additionalConfig[DdSdkReactNative.DD_NATIVE_VIEW_TRACKING_KEY] = configuration.nativeViewTracking;
  configuration.additionalConfig[DdSdkReactNative.DD_NATIVE_INTERACTION_TRACKING_KEY] = configuration.nativeInteractionTracking;

  if (configuration.verbosity) {
    configuration.additionalConfig[DdSdkReactNative.DD_SDK_VERBOSITY_KEY] = configuration.verbosity;
  }

  if (configuration.proxyConfig) {
    const additionalConfig = configuration.additionalConfig;
    const proxyConfig = configuration.proxyConfig;
    additionalConfig[DdSdkReactNative.DD_PROXY_TYPE_KEY] = proxyConfig.type;
    additionalConfig[DdSdkReactNative.DD_PROXY_ADDRESS_KEY] = proxyConfig.address;
    additionalConfig[DdSdkReactNative.DD_PROXY_PORT_KEY] = proxyConfig.port;

    if (proxyConfig.username && proxyConfig.password) {
      if (proxyConfig.type === _ProxyConfiguration.ProxyType.SOCKS) {
        console.warn("SOCKS proxy configuration doesn't support Basic authentication.");
      } else {
        additionalConfig[DdSdkReactNative.DD_PROXY_USERNAME_KEY] = proxyConfig.username;
        additionalConfig[DdSdkReactNative.DD_PROXY_PASSWORD_KEY] = proxyConfig.password;
      }
    }
  }

  if (configuration.serviceName) {
    configuration.additionalConfig[DdSdkReactNative.DD_SERVICE_NAME] = configuration.serviceName;
  }

  if (configuration.version) {
    configuration.additionalConfig[DdSdkReactNative.DD_VERSION] = `${configuration.version}${configuration.versionSuffix ? `-${configuration.versionSuffix}` : ''}`;
  } // If both version and version suffix are provided, we merge them into the version field.
  // To avoid adding it in again the native part, we only set it if the version isn't set.


  if (configuration.versionSuffix && !configuration.version) {
    configuration.additionalConfig[DdSdkReactNative.DD_VERSION_SUFFIX] = `-${configuration.versionSuffix}`;
  }

  configuration.additionalConfig['_dd.first_party_hosts'] = (0, _DdSdkReactNativeConfiguration.formatFirstPartyHosts)(configuration.firstPartyHosts);
});
