// @datadog/mobile-react-native/lib/commonjs/ProxyConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProxyType = exports.ProxyConfiguration = void 0;

/**
 * Proxy server configuration.
 */
class ProxyConfiguration {
  constructor(
  /**
   * Proxy type.
   */
  type,
  /**
   * Proxy address. Can be either in the IP address format, ex. "1.1.1.1", or hostname, ex. "example.com".
   */
  address,
  /**
   * Proxy port.
   */
  port,
  /**
   * Username for Basic authentication scheme.
   *
   * Note: "SOCKS + authentication" scheme is not supported.
   */
  username,
  /**
   * Password for Basic authentication scheme.
   */
  password) {
    this.type = type;
    this.address = address;
    this.port = port;
    this.username = username;
    this.password = password;
  }

}

exports.ProxyConfiguration = ProxyConfiguration;
var ProxyType;
exports.ProxyType = ProxyType;

(function (ProxyType) {
  ProxyType["HTTP"] = "http";
  ProxyType["HTTPS"] = "https";
  ProxyType["SOCKS"] = "socks";
})(ProxyType || (exports.ProxyType = ProxyType = {}));
