// @datadog/mobile-react-native/lib/commonjs/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DdSdkConfiguration = void 0;

/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

/**
 * A configuration object to initialize Datadog's features.
 */
class DdSdkConfiguration {
  constructor(clientToken, env, applicationId, nativeCrashReportEnabled, nativeLongTaskThresholdMs, longTaskThresholdMs, sampleRate, site, trackingConsent, additionalConfig, telemetrySampleRate, vitalsUpdateFrequency, configurationForTelemetry) {
    this.clientToken = clientToken;
    this.env = env;
    this.applicationId = applicationId;
    this.nativeCrashReportEnabled = nativeCrashReportEnabled;
    this.nativeLongTaskThresholdMs = nativeLongTaskThresholdMs;
    this.longTaskThresholdMs = longTaskThresholdMs;
    this.sampleRate = sampleRate;
    this.site = site;
    this.trackingConsent = trackingConsent;
    this.additionalConfig = additionalConfig;
    this.telemetrySampleRate = telemetrySampleRate;
    this.vitalsUpdateFrequency = vitalsUpdateFrequency;
    this.configurationForTelemetry = configurationForTelemetry;
  }

}
/**
 * The entry point to initialize Datadog's features.
 */


exports.DdSdkConfiguration = DdSdkConfiguration;
